foyer/forcefield.py

```diff
@@ -330,6 +340,49 @@ def apply(self, topology, references_file=None, *args, **kwargs):
 
         return structure
 
+    def run_atomtyping(self, topology, use_residue_map=True):
+        """Atomtype the topology
+
+        Parameters
+        ----------
+        topology : openmm.app.Topology
+            Molecular structure to find atom types of
+        use_residue_map : boolean
+            Store atomtyped topologies of residues to a dictionary that maps
+            them to residue names.  Each topology, including atomtypes, will be
+            copied to other residues with the same name. This avoids repeatedly
+            calling the subgraph isomorphism on idential residues and should
+            result in better performance for systems with many identital
+            residues, i.e. a box of water. Note that for this to be applied to
+            independent molecules, they must each be saved as different
+            residues in the topology.
+        """
+        if use_residue_map:
+            independent_residues = _check_independent_residues(topology)
+
+            if independent_residues:
+                residue_map = dict()
+
+                for res in topology.residues():
+                    if res.name not in residue_map.keys():
+                        residue = _topology_from_residue(res)
+                        find_atomtypes(residue, forcefield=self)
+                        residue_map[res.name] = residue
+
+                for key, val in residue_map.items():
+                    _update_atomtypes(topology, key, val)
+
+            else:
+                find_atomtypes(topology, forcefield=self)
+
+        else:
+            find_atomtypes(topology, forcefield=self)
+
+        if not all([a.id for a in topology.atoms()][0]):
+            raise ValueError('Not all atoms in topology have atom types')
+
+        return topology
+
     def createSystem(self, topology, atomtype=True, use_residue_map=True,
```

Comment:
We should be able to remove the `atomtype` and `use_residue_map` arguments now. In fact, is `createSystem` now exactly the same as OpenMM's? If so, we could actually delete this entire function... (although this may cause problems with #155). Thoughts @ctk3b?

Reply:
Double check me but I think the base method includes the logic for template based atomtyping

Reply:
I'll clean up the arguments but this and a few other ideas with `Forcefield.createSystem` vs OpenMM's original function should be dealt with at some point. There's an open issue about templating or something that I can't find at the moment but should be referenced.

Reply:
We think you're right in saying that they do their reside teamplate-based atomtyping in their `Forcefield.createSystem`. It doesn't appear we can use their base method out of the box.

Worth noting that their while templates are much more involved, we are already doing some work in creating a map between residue/molecule names to atomtyped topologies but trashing it once the system is atomtyped. I don't know if it's feasible or worth considering trying to save `residue_map` to disk? Or maybe converting it into a single large `residueTemplate` that could then be passed to OpenMM's base method?

Here's their method at a point in history that should be close to when we grabbed it:

https://github.com/pandegroup/openmm/blob/c9fcabb5da87ef61ac0c777423b1f3e1d96e654e/wrappers/python/simtk/openmm/app/forcefield.py#L1041

Also notable that there have been some updates to that block of in the past year. Probably not much of consequence now but something to worry about later, especially if we want great integration with OpenMM ...